### ISSUE TITLE
Possible bug in `ElementQuery::ids()` method

### ISSUE THREAD

**Issue by putyourlightson** [craftcms/cms#5216]
I'm not sure if this should be classified as a bug in Craft (3.3.x) or Yii, but after investigating [this issue](https://github.com/putyourlightson/craft-blitz/issues/140), I found that executing the following query:
```
\craft\elements\Entry::find()->indexBy('id')->ids();
```

throws an `Integrity constraint violation` error:
```
Integrity constraint violation – yii\db\IntegrityException
SQLSTATE[23000]: Integrity constraint violation: 1052 Column 'id' in field list is ambiguous
The SQL being executed was: SELECT `elements`.`id`, `id`
FROM (SELECT `elements`.`id` AS `elementsId`, `elements_sites`.`id` AS `elementsSitesId`, `content`.`id` AS `contentId`, `structureelements`.`structureId`
FROM `elements` `elements`
INNER JOIN `entries` `entries` ON `entries`.`id` = `elements`.`id`
INNER JOIN `elements_sites` `elements_sites` ON `elements_sites`.`elementId` = `elements`.`id`
INNER JOIN `content` `content` ON (`content`.`elementId` = `elements`.`id`) AND (`content`.`siteId` = `elements_sites`.`siteId`)
LEFT JOIN `structureelements` `structureelements` ON (`structureelements`.`elementId` = `elements`.`id`) AND (EXISTS (SELECT *
FROM `structures`
WHERE (`id` = `structureelements`.`structureId`) AND (`dateDeleted` IS NULL)))
WHERE (`elements_sites`.`siteId`='1') AND (`elements`.`archived`=FALSE) AND (((`elements`.`enabled`=TRUE) AND (`elements_sites`.`enabled`=TRUE)) AND (`entries`.`postDate` <= '2019-11-06 17:22:18') AND ((`entries`.`expiryDate` IS NULL) OR (`entries`.`expiryDate` > '2019-11-06 17:22:18'))) AND (`elements`.`dateDeleted` IS NULL) AND (`elements_sites`.`enabled`=TRUE) AND (`elements`.`draftId` IS NULL) AND (`elements`.`revisionId` IS NULL)
ORDER BY `structureelements`.`lft`, `entries`.`postDate` DESC) `subquery`
INNER JOIN `entries` `entries` ON `entries`.`id` = `subquery`.`elementsId`
INNER JOIN `elements` `elements` ON `elements`.`id` = `subquery`.`elementsId`
INNER JOIN `elements_sites` `elements_sites` ON `elements_sites`.`id` = `subquery`.`elementsSitesId`
INNER JOIN `content` `content` ON `content`.`id` = `subquery`.`contentId`
LEFT JOIN `structureelements` `structureelements` ON (`structureelements`.`elementId` = `subquery`.`elementsId`) AND (`structureelements`.`structureId` = `subquery`.`structureId`)
ORDER BY `structureelements`.`lft`, `entries`.`postDate` DESC
```

The reason for this is that `indexBy` is set to `id`, which is then appended to the select query by Yii on [this line](https://github.com/yiisoft/yii2/blob/2.0.29/framework/db/Query.php#L314): 
```
SELECT `elements`.`id`, `id`
```

Ensuring that `indexBy` is set to `null` in the `ids()` method in `craft\elements\db\ElementQuery` fixes the issue:

    public function ids($db = null): array
    {
        ...

        // Ensure `indexBy` is null to avoid an integrity constraint violation error
        if ($this->indexBy !== null) {
            $this->indexBy(null);
        }

        $select = $this->select;
        $this->select = ['elements.id' => 'elements.id'];
        $result = $this->column($db);

        ...
    }

**Comment by brandonkelly** [comment]
The problem isn’t that the `SELECT` clause contained `id` twice; it’s that `` `id` `` on its isn’t clear _which_ `id` column to select, but multiple tables are involved that have `id` columns (`elements`, `elements_sites`, `content`, etc.).

Turns out that when `yii\db\Query::column()` is adding the `indexBy` value to the `SELECT` clause, it does in fact try to give it a table alias in the process. But due to the nature element queries work—the element query is not built into SQL directly; it creates a whole separate Query object behind the scenes based on its abstracted params—`column()` wasn’t sure which table alias to give it.

The next Craft release fixes this by setting `$this->from = ['elements' => Table::ELEMENTS];` from `craft\elements\db\ElementQuery::init()`, as a pointer for `column()`.

In the meantime, if you really need the `ids()` query results to be indexed by ID, I’d leave `indexBy()` out of it, and just `array_flip()` the results, which will be more performant anyway.

```php
$indexedIds = array_flip(Entry::find()->ids());
```

Or if you really want to keep the keys and values equal to each other (`['5' => '5']`, etc.)`, do this:

```php
$ids = Entry::find()->ids();
$indexedIds = array_combine($ids, $ids);
```

**Comment by putyourlightson** [comment]
Thanks Brandon, I can confirm that your fix works!

I'm not sure what @naboo was using `indexBy` for, but I'm glad to see the `from` tables more explicitly set now.

**Comment by naboo** [comment]
Great to see this fixed. I'm using the `indexBy('id')` like this in my code to get the product entries indexed by ID on return:

```$entries = Entry::find()->section('products')->with(['images'])->where(['elements.id' => $elementIds])->indexBy('id')->status(['live', 'expired'])->all();```

As you can see I'm not using it together with `->ids()`. Should I change my code to "elements.id"?

**Comment by brandonkelly** [comment]
@naboo Hm, the only time `indexBy()` will actually modify the element query is if `column()` is called (or `ids()`, which calls `column()` internally). In your case, `indexBy('id')` would only come into play _after_ the query has been executed, and Craft is preparing the response from the raw query data.

Check your stack trace and make sure you’re even looking at the right line in your code.

**Comment by putyourlightson** [comment]
To clarify, the `ids()` method is being called by Blitz in the context of cache invalidation.

**Comment by brandonkelly** [comment]
Aha, gotcha.

**Comment by naboo** [comment]
So we are all good? :)

Or should I still look into modifying something on my end? This line of
code is the only place I use the function and it's in a custom module and
function which has nothing to do with anything unless it's called.

Den tors 7 nov. 2019 18:12Brandon Kelly <notifications@github.com> skrev:

**Comment by putyourlightson** [comment]
Brandon's commit has resolved the issue for the next release of Craft :)

**Comment by naboo** [comment]
Awesome. Thanks all!

Den tors 7 nov. 2019 18:25Ben Croker <notifications@github.com> skrev:

**Comment by boboldehampsink** [comment]
@brandonkelly this causes a new bug for me.

Running this:

`User::find()->username('demo_bedrijf_*')->indexBy('username')->ids();`

Now results in:

```
Next yii\db\Exception: SQLSTATE[42703]: Undefined column: 7 ERROR:  column elements.username does not exist
web_1         | LINE 1: SELECT "elements"."id", "elements"."username"
web_1         |                                 ^
web_1         | The SQL being executed was: SELECT "elements"."id", "elements"."username"
web_1         | FROM (SELECT "elements"."id" AS "elementsId", "elements_sites"."id" AS "elementsSitesId", "content"."id" AS "contentId"
web_1         | FROM "elements" "elements"
web_1         | INNER JOIN "users" "users" ON "users"."id" = "elements"."id"
web_1         | INNER JOIN "elements_sites" "elements_sites" ON "elements_sites"."elementId" = "elements"."id"
web_1         | INNER JOIN "content" "content" ON ("content"."elementId" = "elements"."id") AND ("content"."siteId" = "elements_sites"."siteId")
web_1         | WHERE ("users"."username" ILIKE 'demo_bedrijf_%') AND ("elements_sites"."siteId"=4) AND ("elements"."archived"=FALSE) AND (("users"."suspended"=FALSE) AND ("users"."pending"=FALSE)) AND ("elements"."dateDeleted" IS NULL) AND ("elements_sites"."enabled"=TRUE) AND ("elements"."draftId" IS NULL) AND ("elements"."revisionId" IS NULL)
web_1         | ORDER BY "users"."username") "subquery"
web_1         | INNER JOIN "users" "users" ON "users"."id" = "subquery"."elementsId"
web_1         | INNER JOIN "elements" "elements" ON "elements"."id" = "subquery"."elementsId"
web_1         | INNER JOIN "elements_sites" "elements_sites" ON "elements_sites"."id" = "subquery"."elementsSitesId"
web_1         | INNER JOIN "content" "content" ON "content"."id" = "subquery"."contentId"
web_1         | ORDER BY "users"."username" in /app/user/vendor/yiisoft/yii2/db/Schema.php:674
```

**Comment by brandonkelly** [comment]
@boboldehampsink Hm, able to reproduce, but nothing we can do in Craft itself unfortunately. Your code was only working because element queries weren’t properly supporting `indexBy` in the first place, which was fixed per this issue.

I have just submitted yiisoft/yii2#17688 which would make this possible:

```php
User::find()
    ->username('demo_bedrijf_*')
    ->indexBy('users.username')
    ->ids();
```

In the meantime you could use `select()` + `pairs()` instead:

```php
User::find()
    ->username('demo_bedrijf_*')
    ->select(['username', 'elements.id'])
    ->pairs();
```

**Comment by boboldehampsink** [comment]
Ok, thanks. But... this isn't backwards compatible, so maybe defer it to 3.4? I now have to scan all my projects for `indexBy` uses before updating to 3.3.16

**Comment by boboldehampsink** [comment]
@brandonkelly this also means that we can now only use `indexBy` with columns from `elements`?

**Comment by brandonkelly** [comment]
@boboldehampsink Sorry, of course you are right. I just released 3.3.16.2 with a better fix for this issue, that doesn’t introduce a BC break.
